src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/GenerateDockerComposeV3.scala

```diff
@@ -0,0 +1,175 @@
+package au.com.agiledigital.toolform.tasks.generate.docker
+
+import java.io.{BufferedWriter, FileWriter}
+
+import au.com.agiledigital.toolform.app.{ToolFormConfiguration, ToolFormError}
+import au.com.agiledigital.toolform.model._
+import au.com.agiledigital.toolform.tasks.generate.{WriterContext, YamlWriter}
+import au.com.agiledigital.toolform.tasks.generate.docker.DockerFormatting._
+import au.com.agiledigital.toolform.tasks.generate.docker.SubEdgeDef.subEdgeDefsFromProject
+import au.com.agiledigital.toolform.version.BuildInfo
+
+import scalaz.Scalaz._
+import scalaz._
+
+class GenerateDockerComposeV3() extends YamlWriter {
+
+  // TODO: Surely Scalaz has a built in identity function that could replace this.
+  // Maybe it could be replaced with a different construct altogether?
+  private val identity = State[WriterContext, Unit] { context =>
+    (context, ())
+  }
+
+  def runDockerComposeV3(toolFormConfiguration: ToolFormConfiguration, project: Project): Either[ToolFormError, String] = {
+    val outFile = toolFormConfiguration.generateTaskConfiguration.out
+    if (outFile.isDirectory) {
+      return Left(ToolFormError("Output path is a directory. Docker Compose V3 output requires a single file as an output."))
+    }
+
+    // Logic
+
+    val writer = new BufferedWriter(new FileWriter(outFile, false))
+    try {
+
+      val writeFile = for {
+        _ <- write(s"# Generated by ${BuildInfo.name} (${BuildInfo.version})")
+        _ <- write("version: '3'")
+        _ <- write("services:")
+        i <- indent()
```

Comment:
```
i <- indent()
_ <- ...
_ <- resetIndent(i)
```

occurs frequently. Consider how to express as something like:
```
_ <- indented {
  writeEdges(...),
  writeComponents(...),
  writeResources(...) 
}
```

Reply:
I have used a construct that looks like this

```
_ <- indented { 
  for {
    _ <- writeEdges(...)
    _ <- writeComponents(...),
    _ <- writeResources(...) 
  } 
}
```

Providing the method calls as parameters to a method as shown in your example could be done and looks a bit neater but would require some plumbing to chain the monads inside the indented function which could add complexity. This way is simpler to implement and maintain and still has the advantage of making the intention clear. 

What do you think?